### PR TITLE
Experiment with holding hard references to objects being qdeleted in 515 (saves 1.1+ seconds on init times, more on prod)

### LINKED ISSUE #10: Migrate SSgarbage's queue from using \ref to using direct references
## What sucks

\ref is real slow, at least in the scale of qdel. This is primarily because adding a getting ref requires inserting a new string into the string tree, and we bloat that sucker to hell.
We use refs for SSgarbage's queue, because it lets us assert that byond's garbage collector will do its work if there are no other refs remaining 

## How fix

515 adds a var that gives us the amount of references an object has remaining. We could leverage this alongside some information we know about how many refs an object "should" have by that point to save the cost of all those \refs

That works out to about 1500ms

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request
Adds `EXPERIMENT_515_QDEL_HARD_REFERENCE`, which will queue to the GC subsystem using hard references rather than `\ref`.  This is only possible in 515 because of the new `refcount` proc. `\ref` is very very slow and has some nasty knock on effects, so removing its usages where possible is good.

This is an explicit opt in define because I want to give us the ability to test 515 on live while only testing 515 itself, not our experimental changes. We have a few more of these we want to do so I made a separate file for them. They're auto-defined in unit tests so we see them with the alternate test runner. In a perfect world we'd test both on and off, but eh.

Closes https://github.com/tgstation/dev-cycles-initiative/issues/10

### CODE REVIEW

File: code/controllers/subsystem/garbage.dm [hunk 1]
```diff
@@ -139,6 +139,13 @@ SUBSYSTEM_DEF(garbage)
 			pass_counts[i] = 0
 			fail_counts[i] = 0
 
+#ifdef EXPERIMENT_515_QDEL_HARD_REFERENCE
+// 1 from the hard reference in the queue, and 1 from the variable used before this
+#define IS_DELETED(datum, _) (refcount(##datum) == 2)
```
Comment (LemonInTheDark) [inline code/controllers/subsystem/garbage.dm:144]: I wish this had better comments as to why we're doing it

Comment (Mothblocks) [reply code/controllers/subsystem/garbage.dm:144]: What do you have in mind?

Comment (LemonInTheDark) [reply code/controllers/subsystem/garbage.dm:144]: I'm unsure, maybe just a blurb about how this should be faster then using \ref constantly. it'd be easier to say if ssgarbage was more thoroughly explained somewhere

**Comment by Mothblocks** [comment]
Tests are failing but because of test specific assumptions

**Comment by Mothblocks** [comment]
I think this caught real bugs in guns?

**Comment by Mothblocks** [comment]
Ok it's not really a bug I'll explain later

**Comment by Mothblocks** [comment]
image

File: code/controllers/subsystem/garbage.dm (hunk 1 above)
Comment (Mothblocks) [reply code/controllers/subsystem/garbage.dm:144]: I'll do it later sure but for the record this macro should be removed when `EXPERIMENT_515_QDEL_HARD_REFERENCE` is removed

**Comment by Mothblocks** [comment]
For posterity, we realized later on we can make this use pointers (basically first class weakrefs), and avoid most of this. I'm just taking it upon myself to classify these as flaky improper cleanup anyway, and we can switch to pointers if it becomes unwieldy.

File: code/modules/pai/pai.dm [hunk 2]
```diff
@@ -159,12 +159,7 @@
 	QDEL_NULL(internal_gps)
 	QDEL_NULL(newscaster)
 	QDEL_NULL(signaler)
-	if(!QDELETED(card) && loc != card)
-		card.forceMove(drop_location())
-		// these are otherwise handled by paicard/handle_atom_del()
-		card.pai = null
-		card.emotion_icon = initial(card.emotion_icon)
-		card.update_appearance()
```
Comment (Mothblocks) [inline code/modules/pai/pai.dm:167]: @jlsnow301 Please advise, this should never have been on Destroy() (forceMove is an observable side effect) so I'm removing this, but not really sure what we're losing from this.

**Comment by Mothblocks** [comment]
Newest bug is stuff reporting that it's qdel'ing before initialize.

This is a bug that has always existed because BYOND makes no GC guarantees but is now made consistent.

It's stuff in preferences creating humans before the atoms SS, which *sometimes* would've been cleared up by BYOND, but because now it definitely won't, it will eventually be initialized by `InitAtom`, which will see it's qdeleted, and fail.

Solution is to not care about stuff that's qdel'd before atoms SS started running.

**Comment by Mothblocks** [comment]
Looking at logs of that error being tripped, nearly all can be suspected to be of prefs menu, or at the very least, very little can be suspected of being pre-atoms init

**Comment by Mothblocks** [comment]
image

File: code/modules/pai/pai.dm (hunk 2 above)
Comment (jlsnow301) [reply code/modules/pai/pai.dm:167]: pais outside of their card are set to drop the card, hopefully that behavior remains, but the rest of this should be okay to delete

**Comment by Mothblocks** [comment]
Nevermind this is the only way to get them to work

image

**Comment by Mothblocks** [comment]
@LemonInTheDark

File: code/modules/pai/pai.dm (hunk 2 above)
Comment (LemonInTheDark) [reply code/modules/pai/pai.dm:167]: @Mothblocks behavior still good? destruction/death maybe?

File: code/controllers/subsystem/garbage.dm [hunk 3]
```diff
@@ -169,16 +176,21 @@ SUBSYSTEM_DEF(garbage)
...
+#endif
 
-		if (!D || D.gc_destroyed != GCd_at_time) // So if something else coincidently gets the same ref, it's not deleted by mistake
+		if (IS_DELETED(D, GCd_at_time)) // So if something else coincidently gets the same ref, it's not deleted by mistake
 			++gcedlasttick
 			++totalgcs
 			pass_counts[level]++
 			#ifdef REFERENCE_TRACKING
-			reference_find_on_fail -= refID //It's deleted we don't care anymore.
+			reference_find_on_fail -= text_ref(D) //It's deleted we don't care anymore.
```
Comment (LemonInTheDark) [inline code/controllers/subsystem/garbage.dm:193]: will add overhead to unit tests btw

**Review by LemonInTheDark — CHANGES_REQUESTED** [review]

**Comment by LemonInTheDark** [comment]
PAI thing fixed?

File: code/controllers/subsystem/garbage.dm (hunk 3 above)
Comment (Mothblocks) [reply code/controllers/subsystem/garbage.dm:193]: 515 tests seem to run faster than the 514 ones :P

**Comment by Mothblocks** [comment]
I'm not really sure what it is still

**Comment by LemonInTheDark** [comment]
if a pai hologram "dies" it should drop its card so the pai can continue to live

**Comment by Mothblocks** [comment]
Let me see what the repro is before/after

File: .github/workflows/ci_suite.yml [hunk 4]
```diff
@@ -111,16 +111,17 @@ jobs:
...
       fail-fast: false
       matrix:
         map: ${{ fromJSON(needs.find_all_maps.outputs.maps).paths }}
     concurrency:
       group: run_all_tests-${{ github.ref }}-${{ matrix.map }}
       cancel-in-progress: true
-    uses: ./.github/workflows/run_integration_tests.yml
-    with:
-      map: ${{ matrix.map }}
+    # uses: ./.github/workflows/run_integration_tests.yml
```
Comment (Mothblocks) [inline .github/workflows/ci_suite.yml]: why is this still here. i thought if ixed this

**Comment by Mothblocks** [comment]
@LemonInTheDark done

File: code/controllers/subsystem/garbage.dm (hunk 3 above)
Comment (LemonInTheDark) [reply code/controllers/subsystem/garbage.dm:193]: ~~I meant the 514 ones~~

**Pull request merged**